contracts/cosmwasm-vm/cw-xcall/src/send_call_message.rs

```diff
@@ -17,47 +22,84 @@ impl<'a> CwCallService<'a> {
         sources: Vec<String>,
         destinations: Vec<String>,
     ) -> Result<Response, ContractError> {
-        let caller = info.sender;
-        let config = self.get_config(deps.as_ref().storage)?;
-        let nid = config.network_id;
-
-        self.ensure_caller_is_contract_and_rollback_is_null(deps.as_ref(), &caller, &rollback)?;
-
-        let need_response = rollback.is_some();
-
-        let rollback_data = match rollback {
-            Some(data) => data,
-            None => vec![],
+        let msg = if rollback.is_some() {
+            AnyMessage::CallMessageWithRollback(CallMessageWithRollback {
+                msg_type: MessageType::MessageWithRollback,
+                data,
+                rollback: rollback.unwrap(),
+            })
+        } else {
+            AnyMessage::CallMessage(CallMessage {
+                data,
+                msg_type: MessageType::BasicMessage,
+            })
         };
+        let envelope = Envelope::new(msg, sources, destinations);
+        self.send_call(deps, info, to, envelope)
+    }
 
-        self.ensure_rollback_length(&rollback_data)?;
-        println!("{LOG_PREFIX} Packet Validated");
+    pub fn validate_payload(
+        &self,
+        deps: Deps,
+        caller: &Addr,
+        envelope: &Envelope,
+    ) -> Result<(), ContractError> {
+        match &envelope.message {
+            AnyMessage::CallMessage(_m) => Ok(()),
+            AnyMessage::CallMessageWithRollback(m) => {
+                if !is_contract(deps.querier, caller) {
+                    return Err(ContractError::RollbackNotPossible);
+                }
+                self.ensure_rollback_length(&m.rollback().unwrap())?;
+                Ok(())
+            }
+        }
+    }
+
+    pub fn send_call(
+        &self,
+        deps: DepsMut,
+        info: MessageInfo,
+        to: NetworkAddress,
+        envelope: Envelope,
+    ) -> Result<Response, ContractError> {
+        let caller = info.sender.clone();
+        let config = self.get_config(deps.as_ref().storage)?;
+        let nid = config.network_id;
+        self.validate_payload(deps.as_ref(), &caller, &envelope)?;
 
         let sequence_no = self.get_next_sn(deps.storage)?;
-        let mut confirmed_sources = sources.clone();
+        let mut confirmed_sources = envelope.sources.clone();
         let from = NetworkAddress::new(&nid, caller.as_ref());
 
         if confirmed_sources.is_empty() {
             let default = self.get_default_connection(deps.as_ref().storage, to.nid())?;
             confirmed_sources = vec![default.to_string()]
         }
 
-        if need_response {
-            let request =
-                CallRequest::new(caller.clone(), to.clone(), sources, rollback_data, false);
+        if envelope.message.rollback().is_some() {
```

Comment:
we can do it both ways ... in flow we can stick with the flags the message types are to simplify the flow to users so that they dont need to worry explicity about the flags.